### PR TITLE
platform/aws: create security groups if they don't exist

### PR DESCRIPTION
**Pull request by euank**
cc @bgilbert @marineam

### CODE REVIEW

File: platform/api/aws/ec2.go [hunk 1]
```diff
@@ -200,3 +204,89 @@ func (a *API) CreateTags(resources []string, tags map[string]string) error {
...
+	}
+	if len(sgIds.SecurityGroups) == 0 {
+		return "", fmt.Errorf("zero security groups matched name %v", name)
+	}
+	return *sgIds.SecurityGroups[0].GroupId, nil
+}
+
+// createSG creates a security group with tcp/22 access allowed from the
+// internet.
+func (a *API) createSG(name string) (string, error) {
```
Comment (bgilbert) [inline platform/api/aws/ec2.go]: `createSecurityGroup`?

File: platform/api/aws/ec2.go [hunk 2]
```diff
@@ -200,3 +204,89 @@ func (a *API) CreateTags(resources []string, tags map[string]string) error {
...
+	sg, err := a.ec2.CreateSecurityGroup(&ec2.CreateSecurityGroupInput{
+		GroupName:   aws.String(name),
+		Description: aws.String("kola security group for testing"),
```
Comment (bgilbert) [inline platform/api/aws/ec2.go]: `mantle`

File: platform/api/aws/ec2.go [hunk 3]
```diff
@@ -200,3 +204,89 @@ func (a *API) CreateTags(resources []string, tags map[string]string) error {
...
+		_, err := a.ec2.AuthorizeSecurityGroupIngress(&input)
+
+		if err != nil {
+			// We created the SG but can't add all the needed rules, let's try to
+			// bail gracefully
+			_, delErr := a.ec2.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+				GroupId: sg.GroupId,
+			})
+			if delErr != nil {
+				return "", fmt.Errorf("created sg %v (%v) but couldn't authorize it. Manual deltion may be required: %v", *sg.GroupId, name, err)
```
Comment (bgilbert) [inline platform/api/aws/ec2.go]: `deletion`

File: platform/api/aws/ec2.go [hunk 4]
```diff
@@ -200,3 +204,89 @@ func (a *API) CreateTags(resources []string, tags map[string]string) error {
...
+			}
+			return "", fmt.Errorf("created sg %v (%v), but couldn't authorize it and thus deleted it: %v", *sg.GroupId, name, err)
+		}
+	}
+	return *sg.GroupId, err
+}
+
+func isSecurityGroupNotExist(err error) bool {
```
Comment (bgilbert) [inline platform/api/aws/ec2.go]: This seems so specific that I'd be tempted to inline it, but I could see it going either way.

**Comment by euank** [comment]
I think for ec2-classic you have to specify the vpc for a few of these calls. Maybe a created-default-vpc will still work correctly, I'm a little fuzzy on all the details.

File: platform/api/aws/ec2.go [hunk 5]
```diff
@@ -200,3 +204,89 @@ func (a *API) CreateTags(resources []string, tags map[string]string) error {
...
+	})
+	if isSecurityGroupNotExist(err) {
+		return a.createSecurityGroup(name)
+	}
+
+	if err != nil {
+		return "", fmt.Errorf("unable to get security group named %v: %v", name, err)
+	}
+	if len(sgIds.SecurityGroups) == 0 {
+		return "", fmt.Errorf("zero security groups matched name %v", name)
```
Comment (marineam) [inline platform/api/aws/ec2.go]: Is it possible for this to happen and err to be nil?

Comment (euank) [inline platform/api/aws/ec2.go]: I don't think so, this is being overtly defensive

Comment (marineam) [inline platform/api/aws/ec2.go]: Ok, if it was normal I would have expected it to call create but being an unexpected case an error is fine.

**Pull request merged**